File: ui/src/domain/app-rules-preset.ts
/**
 * 应用分流预设 —— **前端不再持有表本体**。
 *
 * 内置 16 条的单一真值在 Rust（`crates/config-engine/src/user_config/app_rules_preset_data.rs`），
 * 经 `app_presets_list` command 下发，由 `store/use-app-presets-store.ts` 启动时拉一次入 store。
 * 本模块只剩**跨语言类型镜像**（①类）+ **列表组合/查找**（②类 UI 派生）。
 *
 * 历史（本批终结）：这里曾有 `APP_PRESETS` 全表 + `QURE_BASE`，与 Rust 侧 16 条同构 —— 而 Rust 那份
 * 文件头自认「来源 = 本文件（1:1 提取）」。即**同一份数据两处维护、且两边都以为对方是抄本**。
 * 迁移前已做逐字段对拍取证：16/16 全等（labelKey/emoji/iconUrl/geositeTags/geoipTags/processNames/
 * category），并对 Rust 表做变异验证 6/6 转红 → 确认 Rust 那份是忠实转录而非有损手抄，方才删除本表。
 *
 * `defaultAppRules` / `seedDefaultAppRules` / `getAppPresetsByCategory` 一并删除：Rust 已有
 * （`default_app_rules` / `seed_default_app_rules`），且前端**零消费**（本批 grep 实证）。
 */
import { CustomAppPreset } from '../contracts/types';

export interface AppPreset {
  id: string;
  /** i18n key，对应 rules.apps.XXX；自定义应用则直接是其 name（见 appBuiltin 判定） */
  labelKey: string;
  /** iconUrl 加载失败时的兜底图标 */
  emoji: string;
  /** Qure Color 彩色图标集 */
  iconUrl?: string;
  geositeTags: string[];
  geoipTags?: string[];
  /** 用于跨平台（macOS/Windows/Linux）按进程名精准分流 */
  processNames?: string[];
  category: 'video' | 'social' | 'ai' | 'tools' | 'game';
}

/**
 * 自定义预设 → AppPreset 形态（**单一转换点**）。
 *
 * `category` 取真实值而非占位：原 `getAppPreset` 把自定义应用的 category 硬写 `'tools'`
 * （注释称「后端不消费 category，此处仅占位满足类型」），而 `AppPolicyScreen` 自己内联的那份
 * 转换用的是 `c.category ?? 'tools'`（真值）——**两份实现行为不一致，且内联那份才是对的**
 * （分类筛选/分组要真值）。收口取真值，无消费方回退。
 */
export function customToPreset(c: CustomAppPreset): AppPreset {
  return {
    id: c.id,
    labelKey: c.name, // 自定义应用直接存储名称（非 i18n key）
    emoji: c.emoji,
    iconUrl: c.iconUrl,
    geositeTags: c.geositeTags,
    geoipTags: c.geoipTags,
    // 进程名透传给配置生成（与内置预设同一消费点）；自定义应用亦可按进程名精准分流。
    processNames: c.processNames,
    category: (c.category as AppPreset['category']) ?? 'tools',
  };
}

/**
 * 合并内置 + 自定义为统一列表（**单一 merge 收口**）。
 *
 * 内置优先：id 撞内置的自定义项被丢弃 —— 与 `getAppPreset`「先内置、后自定义」及 Rust
 * `get_app_preset` 的优先级一致（自定义永远影子不了内置 id）。
 *
 * 为何合并在前端而不在 command 里：`builtin` 是启动时拉一次的静态表，`custom` 来自 config、
 * 随 `config:changed` 实时变。若在后端合并，前端那份缓存一加自定义应用就脱节。
 */
export function mergeAppPresets(builtin: AppPreset[], custom?: CustomAppPreset[]): AppPreset[] {
  if (!custom || custom.length === 0) return builtin;
  const builtinIds = new Set(builtin.map((p) => p.id));
  return [...builtin, ...custom.filter((c) => !builtinIds.has(c.id)).map(customToPreset)];
}

/**
 * 根据 appId 查找预设（内置 + 自定义合一）。
 *
 * `builtin` 由调用方从 store 传入（Rust 下发的那份）—— 本模块不再自持表，故不能像原来那样
 * 隐式闭包到模块级常量上。
 */
export function getAppPreset(
  appId: string,
  builtin: AppPreset[],
  customPresets?: CustomAppPreset[]
): AppPreset | undefined {
  const hit = builtin.find((p) => p.id === appId);
  if (hit) return hit;
  const custom = customPresets?.find((p) => p.id === appId);
  return custom ? customToPreset(custom) : undefined;
}

File: crates/config-engine/src/user_config/app_rules_preset.rs
//! 内置应用分流预设 —— **Rust 是单一真值（SoT）**，前端经 `app_presets_list` command 拉取。
//!
//! 表本体见 `app_rules_preset_data.rs`（16 条，行是维护单元）。本模块提供两个投影 + 消费函数：
//! - [`AppPreset`]（`all_presets()`）：路由生成消费的子集，**不含 UI 列**（builder 零污染）。
//! - [`AppPresetDto`]（`all_presets_dto()`）：全列（含 labelKey/emoji/iconUrl），下发前端渲染。
//!
//! 历史：本模块曾是 `src/shared/app-rules-preset.ts` 的手抄投影（TS 为真源）。现已反转，TS 表已删。

#![forbid(unsafe_code)]

use crate::user_config::rule::{AppRule, CustomAppPreset, RuleAction};
use serde::Serialize;
use std::collections::HashSet;

/// 应用分流预设（路由生成消费的子集）。上游 `AppPreset` 后端投影。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AppPreset {
    pub id: String,
    pub geosite_tags: Vec<String>,
    pub geoip_tags: Vec<String>,
    pub process_names: Vec<String>,
    pub category: String,
}

/// 内置预设全列 DTO —— `app_presets_list` command 的载荷，对齐前端 `AppPreset` interface
/// （`ui/src/shared/app-rules-preset.ts`）逐字段。
///
/// **键名契约**：`rename_all = "camelCase"` 产出 `labelKey`/`iconUrl`/`geositeTags`/`geoipTags`/
/// `processNames` —— 与 TS interface 一致。改键名 = 破坏前端渲染，且 **tsc 抓不到**（invoke 返回值
/// 是 as-cast）→ 由 `tests/frontend_sot_guard.rs` 锁死。
///
/// `iconUrl` 用 `Option` + `skip_serializing_if`：对齐 TS `iconUrl?: string`（自定义预设可无图标）。
/// `geoipTags`/`processNames` 恒序列化为数组（TS 侧标 `?` 但消费点全是 `|| []` / `.some()`，
/// 空数组与缺省等价，恒发更简单）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppPresetDto {
    pub id: String,
    pub label_key: String,
    pub emoji: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url: Option<String>,
    pub geosite_tags: Vec<String>,
    pub geoip_tags: Vec<String>,
    pub process_names: Vec<String>,
    pub category: String,
}

include!("app_rules_preset_data.rs");

/// 根据 appId 查找预设（先内置，后自定义）。上游 `getAppPreset`。
///
/// 内置预设查 `all_presets()`；自定义预设查 `custom_presets`，将 `CustomAppPreset` 转为兼容格式。
/// 找不到返回 `None`。
pub fn get_app_preset(app_id: &str, custom_presets: &[CustomAppPreset]) -> Option<AppPreset> {
    let builtin = all_presets();
    if let Some(p) = builtin.iter().find(|p| p.id == app_id) {
        return Some(p.clone());
    }
    custom_presets
        .iter()
        .find(|p| p.id == app_id)
        .map(|c| AppPreset {
            id: c.id.clone(),
            geosite_tags: c.geosite_tags.clone(),
            geoip_tags: c.geoip_tags.clone(),
            process_names: c.process_names.clone().unwrap_or_default(),
            // 后端不消费 category；分组呈现由渲染层直接读 custom.category。
            category: "tools".to_string(),
        })
}

/// 按 appId 查全列预设（先内置，后自定义）——[`get_app_preset`] 的 UI 列版本。
///
/// 与 [`get_app_preset`] 的唯一差别是**多带 UI 列**（labelKey/emoji/iconUrl）。消费方：
/// `rule_resource_refs::enumerate_resource_refs` 要 `labelKey` 当引用徽标文案。
///
/// 自定义预设的 `labelKey` 取 `name`（自定义应用直接存名称，非 i18n key —— 渲染端据
/// `RuleResourceRef.appBuiltin` 决定要不要过 i18n）；`category` 取**真实值**而非
/// [`get_app_preset`] 那个 `"tools"` 占位 —— 后者的占位注释「后端不消费 category」在后端成立，
/// 但本 DTO 是给渲染端的，分组呈现要真值。
pub fn get_app_preset_dto(
    app_id: &str,
    custom_presets: &[CustomAppPreset],
) -> Option<AppPresetDto> {
    if let Some(p) = all_presets_dto().into_iter().find(|p| p.id == app_id) {
        return Some(p);
    }
    custom_presets
        .iter()
        .find(|p| p.id == app_id)
        .map(|c| AppPresetDto {
            id: c.id.clone(),
            label_key: c.name.clone(),
            emoji: c.emoji.clone(),
            icon_url: c.icon_url.clone(),
            geosite_tags: c.geosite_tags.clone(),
            geoip_tags: c.geoip_tags.clone(),
            process_names: c.process_names.clone().unwrap_or_default(),
            category: c.category.clone().unwrap_or_else(|| "tools".to_string()),
        })
}

/// 默认应用分流规则：为每个内置预设生成「代理·跟全局」规则。上游 `defaultAppRules`。
pub fn default_app_rules() -> Vec<AppRule> {
    all_presets()
        .iter()
        .map(|p| AppRule {
            app_id: p.id.clone(),
            action: RuleAction::Proxy,
            enabled: true,
            target_server_id: None,
        })
        .collect()
}

/// 一次性默认注入合并（幂等）。上游 `seedDefaultAppRules`。
///
/// 为未配置的预设补默认规则；剔除已下线预设的残留规则；保留用户已配置的预设规则与自定义 app（custom-*）。
pub fn seed_default_app_rules(existing: &[AppRule]) -> Vec<AppRule> {
    let presets = all_presets();
    let valid_ids: HashSet<String> = presets.iter().map(|p| p.id.clone()).collect();
    let kept: Vec<AppRule> = existing
        .iter()
        .filter(|r| valid_ids.contains(&r.app_id) || r.app_id.starts_with("custom-"))
        .cloned()
        .collect();
    let have: HashSet<String> = kept.iter().map(|r| r.app_id.clone()).collect();
    let mut result = kept;
    for r in default_app_rules() {
        if !have.contains(&r.app_id) {
            result.push(r);
        }
    }
    result
}

#[cfg(test)]
mod tests;
